Create templates when creating handlers/actions

weierophinney opened the pull request:
This patch updates the `handler:create` and `action:create` commands with additional functionality when they detect that a `TemplateRendererInterface` service is registered. When that happens, they do the following:

- Use a different skeleton, one that accepts a `TemplateRendererInterface` to the constructor, and generates a zend-diactoros response using the results of rendering a template. The template rendered is named after the top-level namespace and the class name by default (stripping out any `Action` or `Handler` suffix).
- Generates a template file based on looking up a corresponding path for the top-level namespace in the application configuration, and using a suffix based on the renderer type.

This behavior can be altered when running the command:

- The `--without-template` option causes the command to generate an empty handler, with no template awareness.
- The `--with-template-namespace` option allows providing an alternate template namespace to use.
- The `--with-template-name` option allows providing an alternate template name to use (minus the namespace).
- The `--with-template-extension` option allows providing an extension in cases where no `templates.extension` configuration is available and the template renderer is unrecognized, or if you just want to provide an alternate file extension for the generated template.

If no `TemplateRendererInterface` is detected in the container, behavior is as it was previously, and the above options are not exposed.

Fixes #50.

weierophinney committed:
Implements functionality for generating templates for handlers

Uses the handler name in order to determine the top-level namespace and
qualified class name in order to generate a template file. Both the
namespace and template name are normalized to lowercase, dash-separated
words.

The class uses two strategies to determine where to put the template:

- If template path configuration exists for the namespace, and only one
  path is present for it, that path is used; if zero or more than one
  are found, it raises an exception, as it cannot decide for the user
  where to put it.
- Otherwise, it determines if the class is in a module, or a flat
  hierarchy. For modules, it puts the template directly beneath the
  module `templates/` directory. Otherwise, it puts it in the
  `templates/{namespace}/` directory off the project root.

weierophinney committed:
Return a value object instead of just the path

We will likely want the template name as well as the path. The command
will use the template name when generating contents for the newly
generated handler, and report the path to the user.

weierophinney committed:
Extract normalizeTemplateIdentifier to a trait

weierophinney committed:
Extract more methods to a trait

Renames the NormalizeTemplateIdentifierTrait to TemplateResolutionTrait,
and adds a number of new methods:

- getNamespace
- getTemplateNameFromClass
- getClassName
- getTemplateNameFromClass
- getConfig

Each of these will be useful in the command class as well.

The patch also simplifies detection of the root namespace, using the
first namespace segment always.

weierophinney committed:
Use container to pull config and renderer dependencies

This ensures that we resolve the renderer correctly, and simplifies
testing against configuration. Additionally, it paves the way for
determining which flags (`--with-template`, `--without-template`) to
expose by default, as we can check for the renderer service.

weierophinney committed:
Ready command to accept template options

This patch updates the `CreateHandlerCommand` to do the following:

- During instantiation, it pulls the application container, and
  determines if a template renderer is registered.

- During `configure()`, if a template renderer is registered, it adds
  the following options:

  - `--without-template`, which will allow disabling generation of a
    template for the created handler.
  - `--with-template-namespace`, which allows overriding the template
    namespace used when generating the template.
  - `--with-template-name`, which allows overriding the template
    name used when generating the template.

- During `execute()`, it varies which class skeleton template to use
  based on whether or not a template renderer is registered and/or the
  `--without-template` flag is enabled. Once generation of the handler
  is done, if a template should be rendered, it is.

I had to do a bit of refactoring to make this work.

The `CreateTemplate` class renames the `forHandlerWithTemplateNamespace`
method to `generateTemplate`, and accepts an additional argument,
`$templateName`. `forHandler()` was modified to detect the template name
and pass that on to `generateTemplate()`.

The `CreateHandler` class was modified to accept the project root and
skeleton template to the constructor. The `generate()` method now
accepts the handler class name, and an optional array of additional
template substitutions, allowing the ability to generate arbitrary class
skeletons.

The `CreateHandlerCommand` class was updated to allow accepting both the
project root and the container to the constructor. This allows easier
testing capabilities. It was also updated to instantiate the
`CreateHandler` class with the new arguments, and to invoke `generate()`
using its new signature.

weierophinney committed:
Updates script file to conditionally include commands

Some commands require things such as the application container file to
be in place. In such cases, the script fails due to autoloading and/or
missing container.

The script now does the following:

- Checks to see if the PWD contains a vendor directory. If so, it does a
  `chdir(getcwd()`, and then a `require 'vendor/autoload.php`.
- It creates a default list of always-available commands.
- If a `config/container.php` file exists, it adds the `handler:create`
  and `action:create` commands.

weierophinney committed:
Better variable replacement strategy

Relying on array order leads to issues when performing template
substitutions using `str_replace()`. This new approach ensures that each
substitution correlates exactly with its variable name.

weierophinney committed:
Re-formats long line

weierophinney commented:
I have a few questions for reviewers:

- Currently, I throw an exception if the renderer type is not in a whitelist. Since this is primarily used to determine what the template suffix is, I have a few other possible approaches:
  - Allow specifying an extension via an option: `--with-template-suffix`.
  - Check for `templates.extension` configuration, and use that. While not all implementations allow that to be configurable, I suspect most do.
  - Fallback to "phtml" or "html". This could be even be done with either or both of the above.

- Currently, if more than one path exists for a given template namespace in the configuration, I raise an exception. I could instead:
  - Always use the first (easy to achieve).
  - Somehow prompt the user to select one (harder to achieve, but possible).

- Currently, if the template namespace does not exist in the paths configuration, I will attempt to determine it based on the generated handler's root namespace. Should I?

Pinging @akrabat, @adamculp, @xtreamwayz, and @webimpress to review.

akrabat commented:
> - Currently, I throw an exception if the renderer type is not in a whitelist. Since this is primarily used to determine what the template suffix is, I have a few other possible approaches:
>   - Allow specifying an extension via an option: `--with-template-suffix`.
>   - Check for `templates.extension` configuration, and use that. While not all implementations allow that to be configurable, I suspect most do.
>   - Fallback to "phtml" or "html". This could be even be done with either or both of the above.

I would like to see at least the second option . If I created a new template renderer, I could provided a `templates.extension` and my renderer becomes a first class citizen.

The first option is a useful bonus for people who insist on using a non-default extension, but frankly there aren't many of them!

I'd prefer not to have a fallback as it's just a guess



> - Currently, if more than one path exists for a given template namespace in the configuration, I raise an exception. I could instead:
>   - Always use the first (easy to achieve).
>   - Somehow prompt the user to select one (harder to achieve, but possible).

At least the first one. Second one would be one of those "surprise and delight the user" situations. I bet nearly noone has more than one path for templates per namespace.


> - Currently, if the template namespace does not exist in the paths configuration, I will attempt to determine it based on the generated handler's root namespace. Should I?

I think so, yes. It's nearly always going to be right.

weierophinney committed:
If more than one template path exists, use the first

Instead of raising an exception.

weierophinney commented:
Thanks for the feedback, @akrabat! With that in place, here is my TODO:

- [x] Add `--with-template-suffix` option support.
- [x] Add support for `templates.extension` configuration when template renderer type is unknown.
- [x] If more than one template path is discovered, use the first.

weierophinney committed:
Allow configurable template extensions

We already supported the `templates.extension` directive, if present.
This patch makes the following changes:

- If the configuration is not present, and the renderer type is not in
  the whitelist, we now raise an exception detailing how the user may
  specify an extension.
- We provide the option `--with-template-extension` for specifying an
  extension via the CLI.
